### PR TITLE
Support user provided babel configs

### PR DESCRIPTION
**Pull request by benstepp**
This change allows the user to provide a babelrc or a babel section in
their package.json. It also allows gatsby to be used without having to
specifiy a babelrc if the user doesn't need anything past what babel
provides as a default.

* It resolves all of the paths to become absolute such that the user can
crawl upwards as in #239.

* It allows the custom usage of other babel plugins so that decorators
work as in #129.

### CODE REVIEW

**Comment by benstepp** [comment]
Alright, I'm fairly happy with how this turned out. Some highlights:

* User can override the `.babelrc` passed to webpack.
* User can use whatever babel plugins they want.
* User babel config is extended with `react-hmre` rather than being overwritten in develop.js.
* Object.assign is now polyfilled by default.
* Non breaking change as starters previously required a babelrc and gatsby will just read them as normal.
* Fixes three issues #129, #235, #264

**Comment by KyleAMathews** [comment]
This is impressive stuff! Way better than what I kludged together! I really like how you've emulated the normal Babel behavior for users. That's exactly how I envision Gatsby working — magic out-of-box awesomesauce experience but can override/extend it just like you would a normal React/Webpack app.

I'll have some more time tomorrow evening to thoroughly review this + merge and cut new release.

File: lib/utils/babel-config.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+function resolvePlugin (pluginName, directory, type) {
+  const gatsbyPath = path.resolve('..', '..')
+  const plugin = resolve(`babel-${type}-${pluginName}`, directory) ||
+    resolve(`babel-${type}-${pluginName}`, gatsbyPath) ||
+    resolve(pluginName, directory) ||
+    resolve(pluginName, gatsbyPath)
+
+  const name = !startsWith(pluginName, 'babel') ? pluginName : `babel-${type}-${pluginName}`
+  const pluginInvariantMessage = `
+  You are trying to use a babel plugin which gatsby cannot find. You
```
Comment (KyleAMathews) [inline lib/utils/babel-config.js]: Capitalize Gatsby here and below.

File: lib/utils/babel-config.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  const plugins = config.plugins || []
+  plugins.forEach(plugin => {
+    normalizedConfig.plugins.push(resolvePlugin(plugin, directory, 'plugin'))
+  })
+
+  return normalizedConfig
+}
+
+/**
+ * Locates a babelrc in the gatsby site root directory. Parses it using
```
Comment (KyleAMathews) [inline lib/utils/babel-config.js]: Capitalize

**Comment by KyleAMathews** [comment]
looking good!

**Comment by KyleAMathews** [comment]
Merged! Will release 0.11 in the morning.

**Pull request merged**